File: Engine/Engine.Shared/BoardCharacteristic.cpp
// ReSharper disable CppCStyleCast
// ReSharper disable CppLocalVariableMayBeConst
// ReSharper disable CppParameterMayBeConst
#include "BoardCharacteristic.h"
#include <vector>
#include <algorithm>
#include "Rule.h"
#include "Utils.h"
#include "InformationFromAuction.h"

BoardCharacteristic::BoardCharacteristic(HandCharacteristic hand, const std::string& previousBidding, const InformationFromAuction& informationFromAuction)
{
    auto bidIds = Utils::SplitAuction(previousBidding);
    position = (int)bidIds.size() + 1;

    partnersSuits = informationFromAuction.partnersSuits;
    opponentsSuit = GetLongestSuit(informationFromAuction.openersSuits);
    stopInOpponentsSuit = GetHasStopInOpponentsSuit(hand.hand, opponentsSuit);

    std::vector<int> suitLengthCombined;
    std::transform(hand.suitLengths.begin(), hand.suitLengths.end(), partnersSuits.begin(), std::back_inserter(suitLengthCombined),
        [](const auto& x, const auto& y) {return x + y; });
    auto firstSuitWithFitIter = std::find_if(suitLengthCombined.begin(), suitLengthCombined.end(), 
        [](const auto& x) {return x >= 8; });
    fitWithPartnerSuit = firstSuitWithFitIter == suitLengthCombined.end() ? -1 : (int)std::distance(suitLengthCombined.begin(), firstSuitWithFitIter);
    hasFit = firstSuitWithFitIter != suitLengthCombined.end();
    fitIsMajor = suitLengthCombined.at(0) >= 8 || suitLengthCombined.at(1) >= 8;

    auto suits = Utils::Split<char>(hand.hand, ',');
    auto trumpSuit = fitWithPartnerSuit == -1 ? "" : suits.at(fitWithPartnerSuit);
    keyCards = Utils::NumberOfCards(hand.hand, 'A') + Utils::NumberOfCards(trumpSuit, 'K');
    trumpQueen = Utils::NumberOfCards(trumpSuit, 'Q');
    lastBidId = Utils::GetLastBidIdFromAuction(previousBidding);

    isCompetitive = Utils::GetIsCompetitive(previousBidding);
    minHcpPartner = informationFromAuction.minHcpPartner;
    allControlsPresent = GetAllControlsPresent(hand, informationFromAuction, fitWithPartnerSuit);
}

int BoardCharacteristic::GetLongestSuit(const std::vector<int>& suitLengths)
{
    return !std::any_of(suitLengths.begin(), suitLengths.end(), [](const auto& x) {return x > 0; }) ? -1 :
        (int)std::distance(suitLengths.begin(), std::max_element(suitLengths.begin(), suitLengths.end()));
}

bool BoardCharacteristic::GetHasStopInOpponentsSuit(const std::string& hand, int opponentsSuit)
{
    if (opponentsSuit == -1)
        return false;
    auto cardsInOpponentSuit = Utils::Split<char>(hand, ',')[opponentsSuit];
    if (cardsInOpponentSuit.length() == 0)
        return false;

    switch (cardsInOpponentSuit[0])
    {
    case 'A': return true;
    case 'K': return cardsInOpponentSuit.length() >= 2;
    case 'Q': return cardsInOpponentSuit.length() >= 3;
    case 'J': return cardsInOpponentSuit.length() >= 4;
    default:
        return false;
    }
}

bool BoardCharacteristic::GetAllControlsPresent(const HandCharacteristic& handCharacteristic, const InformationFromAuction& informationFromAuction, int fitWithPartnerSuit)
{
    std::vector<bool> controlsCombined;
    std::copy(informationFromAuction.controls.begin(), informationFromAuction.controls.end(), std::back_inserter(controlsCombined));
    std::transform(controlsCombined.begin(), controlsCombined.end(), handCharacteristic.controls.begin(), controlsCombined.begin(), [](auto a, auto b) {return a || b;  });
    for (int i = 0; i < 3; i++)
    {
        if (!controlsCombined.at(i) && i != fitWithPartnerSuit)
            return false;
    }
    return true;
}

File: Engine.DotNet/BoardCharacteristic.cs
namespace Engine.DotNet;

public sealed class BoardCharacteristic
{
    public bool HasFit { get; private set; }
    public bool FitIsMajor { get; private set; }
    private List<int> PartnersSuits { get; }
    public int FitWithPartnerSuit { get; }
    public int OpponentsSuit { get; }
    public bool StopInOpponentsSuit { get; private set; }
    public int KeyCards { get; private set; }
    public bool TrumpQueen { get; private set; }
    public int Position { get; private set; }
    public int LastBidId { get; private set; }
    public bool IsCompetitive { get; private set; }
    public int MinHcpPartner { get; private set; }
    public bool AllControlsPresent { get; private set; }

    public BoardCharacteristic(HandCharacteristic hand, string previousBidding, InformationFromAuction informationFromAuction)
    {
        var bidIds = Utils.SplitAuction(previousBidding);
        Position = bidIds.Count + 1;

        PartnersSuits = informationFromAuction.PartnersSuits.ToList();
        OpponentsSuit = GetLongestSuit(informationFromAuction.OpenersSuits);
        StopInOpponentsSuit = GetHasStopInOpponentsSuit(hand.Hand, OpponentsSuit);

        var suitLengthCombined = hand.SuitLengths
            .Zip(PartnersSuits, (x, y) => x + y)
            .ToList();

        var firstSuitWithFitIndex = suitLengthCombined.FindIndex(x => x >= 8);
        FitWithPartnerSuit = firstSuitWithFitIndex == -1 ? -1 : firstSuitWithFitIndex;
        HasFit = firstSuitWithFitIndex != -1;
        FitIsMajor = suitLengthCombined[0] >= 8 || suitLengthCombined[1] >= 8;

        var suits = hand.Hand.Split(",");
        var trumpSuit = FitWithPartnerSuit == -1 ? string.Empty : suits[FitWithPartnerSuit];
        KeyCards = Utils.NumberOfCards(hand.Hand, 'A') + Utils.NumberOfCards(trumpSuit, 'K');
        TrumpQueen = Utils.NumberOfCards(trumpSuit, 'Q') > 0;
        LastBidId = Utils.GetLastBidIdFromAuction(previousBidding);

        IsCompetitive = Utils.GetIsCompetitive(previousBidding);
        MinHcpPartner = informationFromAuction.MinHcpPartner;
        AllControlsPresent = GetAllControlsPresent(hand, informationFromAuction, FitWithPartnerSuit);
    }

    private static int GetLongestSuit(List<int> suitLengths)
    {
        return suitLengths.Any(x => x > 0)
            ? suitLengths.IndexOf(suitLengths.Max())
            : -1;
    }

    private static bool GetHasStopInOpponentsSuit(string hand, int opponentsSuit)
    {
        if (opponentsSuit == -1)
            return false;

        var cardsInOpponentSuit = hand.Split(',')[opponentsSuit];
        if (cardsInOpponentSuit.Length == 0)
            return false;

        return cardsInOpponentSuit[0] switch
        {
            'A' => true,
            'K' => cardsInOpponentSuit.Length >= 2,
            'Q' => cardsInOpponentSuit.Length >= 3,
            'J' => cardsInOpponentSuit.Length >= 4,
            _ => false
        };
    }

    private static bool GetAllControlsPresent(HandCharacteristic handCharacteristic, InformationFromAuction informationFromAuction, int fitWithPartnerSuit)
    {
        var controlsCombined = informationFromAuction.Controls.ToList();
        for (var i = 0; i < 4; i++)
            controlsCombined[i] = controlsCombined[i] || handCharacteristic.Controls[i];

        for (var i = 0; i < 3; i++)
        {
            if (!controlsCombined[i] && i != fitWithPartnerSuit)
                return false;
        }

        return true;
    }
}